### PR TITLE
Quickstart of image signing and enable verification

### PR DESCRIPTION
**Pull request by Xynnn007**
#### Change Log

- Added guide to sign an image with cosign.
- Added guide to configure `policy.json` to enable signature verification to specific image repository.
- Added guide to add cosign verification key and policy to Offline-Fs-KBC's resources.

### CODE REVIEW

**Comment by Xynnn007** [comment]
Hi @sameo @fitzthum @stevenhorsman could you give some suggestions about this? thanks!

File: quickstart.md [hunk 1]
```diff
@@ -831,6 +831,138 @@ kubectl apply -f encrypted-image-tests.yaml
 
 Testing the SEV encrypted container launch can be completed by returning to the section on how to [Verify SEV Memory Encryption](#verify-sev-memory-encryption).
 
+## Signing the encrypted/unencrypted image and enable signature verification when running the workload
+
+When an image is from a protected registry, the image needs to be applied some defined security strategies.
+The strategy types include the following:
```
Comment (stevenhorsman) [inline quickstart.md:837]:
```suggestion
When an image is being pulled from a container registry, [policy requirements](https://github.com/containers/image/blob/main/docs/containers-policy.json.5.md#policy-requirements) can determined whether the image can be accepted.
The requirements can be:
```

**Review by stevenhorsman — COMMENTED** [review]
So, firstly - thank you very much for creating this - it's an area that we are lacking documentation for. I need to go through this more closely and set-up the cosign, but at a high-leve I'm wondering if the quickstart.md isn't the correct place for the docs at the moment (maybe somewhere in image-rs, or a separate doc page?). This guide seems to be talking about how to get cosign support up and running at quite a low level - running the attestation-agent with offline_fs_kbc manually, whereas I think the overall quickstart guide is aimed at users of the installed version of CoCo using the operator.

I think we have some work to do (maybe covered in https://github.com/confidential-containers/community/issues/63) to help enable users to modify their guest images if required (such as when using offline_fs_kbc to test out features like signatures, encrypted images without remote attestation/on non-CC hardware) and some of this information will be great for this, but I'm not sure we are ready for it yet. @fitzthum, @sameo - please feel free to overrule me, if I've missed the point of this quickstart guide

**Review by fitzthum — COMMENTED** [review]
I think we are planning to split up the quickstart guide into several smaller files. It's already too long and too complicated. It would probably be a good idea to make this a separate file and just add a link to it in the guide.

Otherwise I think this looks good. I agree with @stevenhorsman that we really want to focus on deployments that use the operator. I think this is a good starting point, but hopefully we will be able to add some context for simpler deployments. We should have signature support for SEV with the simple-kbs soon so that should help the story.

It might be good to document the kernel parameter that enables signature verification. I don't see that here.

**Comment by Xynnn007** [comment]
> I think we have some work to do (maybe covered in [confidential-containers/community#63](https://github.com/confidential-containers/community/issues/63)) to help enable users to modify their guest images if required (such as when using offline_fs_kbc to test out features like signatures, encrypted images without remote attestation/on non-CC hardware) and some of this information will be great for this, but I'm not sure we are ready for it yet. 

Thanks for you reply Steve! I've checked the link and I'm quite not sure about that. Are you referring that if we add new signatures, it is hard for a user to produce a new guest image which contains the signature?

If so, cosign does not need the image to be modified, because its signatures are stored in the registry where the image lies. 


> I need to go through this more closely and set-up the cosign, but at a high-leve I'm wondering if the quickstart.md isn't the correct place for the docs at the moment (maybe somewhere in image-rs, or a separate doc page?).

Yes I agree with that. I also found that this doc is too long for me to start when editting this.

**Comment by stevenhorsman** [comment]
> Thanks for you reply @stevenhorsman ! I've checked the link and I'm quite not sure about that. Are you referring that if we add new signatures, it is hard for a user to produce a new guest image which contains the signature?

Yes - exactly this point that we don't have a good method for updating the guest image, which the link talks about adding for the sample (option 4), which we could re-use to modify the file. The good news is that with the extra kernel_params exposed in CoCo 0.2 this is becoming a lot easier and the agent-config.toml isn't required now.

> If so, cosign does not need the image to be modified, because its signatures are stored in the registry where the image lies.

So my understanding (which might be wrong) is that with cosign you still need to provide the policy and cosign key via the get_resources endpoint. For remote attestation I think we now don't need to modify the image for this, but the doc made references to offline_fs_kbc, which means we need to modify the guest image to add the `aa-offline_fs_kbc-resources.json` in. I hope that helps clarify?

**Comment by Alex-Carter01** [comment]
Thank you for writing all of this out, I have found it incredibly helpful! 
I have been consulting it enough that I figured I should leave some fixes for typos and such.

**Comment by Xynnn007** [comment]
> So my understanding (which might be wrong) is that with cosign you still need to provide the policy and cosign key via the get_resources endpoint. For remote attestation I think we now don't need to modify the image for this, but the doc made references to offline_fs_kbc, which means we need to modify the guest image to add the `aa-offline_fs_kbc-resources.json` in. I hope that helps clarify?

For this disadvantage, I think there may be still a way to avoid modify guest image frequently.
- Use verdictd-eaa instead of offline-fs-kbc until we finally turn to generic KBS&CC-KBC.
- Use cosign instead of simple signing

cc @arronwy

**Comment by Xynnn007** [comment]
I've put a pr in `image-rs` for this. https://github.com/confidential-containers/image-rs/pull/87  
When it passes, I will change this pr to a single link

File: quickstart.md [hunk 2]
```diff
@@ -920,6 +920,10 @@ kubectl apply -f encrypted-image-tests.yaml
 
 Testing the SEV encrypted container launch can be completed by returning to the section on how to [Verify SEV Memory Encryption](#verify-sev-memory-encryption).
 
+## Signing the encrypted/unencrypted image and enable signature verification when running the workload
+
+Please refer to <https://github.com/confidential-containers/image-rs/blob/main/docs/quickstart/image_signing.md>
+
```
Comment (stevenhorsman) [inline quickstart.md:926]: So I wonder if we need to be a bit more careful in what we are implying here with 'enable signature verification when running the workload'. IIUC, the image-rs doc explains how to sign the image and using image-rs validation that as an image-rs integration test level piece, which is good information, but in order to get signature verification support for the kubernetes workload with CoCo we'd need to do the offline_fs_kbc set-up actions (e.g. passing in the policy, and cosign pub) in the kata/CoCo guest images and we don't have any documentation of how to do that yet, so I wonder if we need to caveats on this doc until we have this?

Comment (Xynnn007) [reply quickstart.md:926]: Yes, you're right. As you've mentioned we do not have a doc to help set up an offline-fs-kbc, maybe still leaving this PR alone until related doc is provided would be a right thing to do. How do you like?

**Pull request closed without merging**